docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,354 @@
+---
+title: "Exported TypeScript Clients Across Wing Applications"
+description: Exported TypeScript Clients Across Wing Applications
+---
+
+# Exported TypeScript Clients Across Wing Applications
+
+- **Author(s)**: @MarkMcCulloh
+- **Submission Date**: 2024-06-02
+
+## Use Case
+
+Given some infrastructure defined in wing by platform engineers, we want to provide a way for application developers in the same organization to interact with that infrastructure in a type-safe way from their own codebases without dealing with wing or infrastructure themselves.
+
+As a platform engineer I'd like to use wing to create the preflight infrastructure for developers in my organization, and then produce client libraries in TypeScript for accessing that infrastructure within other pieces of code.
+Imagine for example that the application developers write TypeScript running in containers in their own repo, and that repo's code is already being automatically deployed with GitOps via ArgoCD or other means.
+
+As an application developer I'd like to consume the TypeScript package published by the platform team and write application code that interacts with these resources (using whichever methods or permissions have been granted by the platform team). If I'm running my TypeScript code / container locally I expect the wing clients to connect to the wing simulator and interact with the simulated resources, and when my code is deployed to the AWS I expect the clients to connect to the real cloud resources.
+
+## Proposal
+
+The `typescript` module has a `Client` class used to create a client. When this is instantiated, `wing compile` will produce a publishable JavaScript npm package with TypeScript type declarations.
+
+```wing
+bring clients;
+
+let client = new typescript.Client(
+  name: "@acme/infra-client",
+  // version: "0.0.0",
+  // outdir: join(@dirname, "dist"),
+);
+```
+
+To lift a resource into the client to make it accessible to consumers, add a `@lift` expression:
+
+```wing
+bring cloud;
+
+let bucket = new cloud.Bucket();
+
+let client = new typescript.Client(
+  name: "@acme/infra-client",
+  lifts: @lift({ bucket: [put] }),
+);
+```
+
+In order to access the lifted objects at runtime, the TypeScript client needs to know how it should obtain the initialization data for instantiating the inflight clients (e.g. resource ARNs and other preflight state). This information is stored in client.context in preflight.
+When the client is used later (inflight), the environment variable `WING_CONTEXT` is expected to the same value previously stored in client.context.
+
+The context is just a string value (base64 encoded JSON), so you can store it in any way that is convenient for your deployment process.
+
+For example, it could be stored in a bucket:
+
+```wing
+bring cloud;
+bring clients;
+bring fs;
+
+let client = new typescript.Client(
+  name: "@acme/infra-client",
+);
```

Comment:
I wish we could set the default node id to be based on `name` because if there are multiple clients, it will look like:

```js
new typescript.Client(name: "foo") as "foo";
new typescript.Client(name: "bar") as "bar";
```

This is a common issue... @Chriscbr I believe we discussed this in the context of revisiting the `as X` syntax. Maybe we can at least allow classes to determine the default node ID?